src/verify_cert.rs

```diff
@@ -205,6 +205,11 @@ fn build_chain_inner(
 
     let err = match result {
         Ok(()) => return Ok(()),
+        // Fatal errors should halt further path building.
```

Comment:
So if the failure from the earlier PR is that we failed to copy the "eagerly return fatal errors" logic, maybe we should extract this `match` bit out in a separate function so that we can reuse it between the two call sites, forcing them to have the same behavior? In other words, maybe moving the `is_fatal()` logic a bit higher up stack?

Reply:
I couldn't think of a good way to express the early return in a simple helper function while also letting the caller get at non-fatal errors. Inspired by [how Brian handled](https://github.com/briansmith/webpki/blob/4a71d475e6da17d4f6fdb60d51cbe6f93a1bbab2/src/verify_cert.rs#L47-L80) similar logic w/ the non-exhaustive `Error` enum in `brainsmith/webpki` I tried implementing an `enum` to express the notion of fatal path building errors vs non-fatal pathbuilding errors in the type system.

@djc Does [this commit](https://github.com/cpu/webpki/commit/c6297e8da87d412458efee7a34bee89baec1cab6) look like an improvement to you? It's a bit more of an invasive diff so I wanted to hold pushing it into this branch until confirming that I'm on the right track.

Reply:
Hmm. I'm not sure I love those suggestions. What about `Continue(Error)` and `Fatal(Error)`. If we're going to avoid the use of "Fatal" we should consider renaming `loop_while_non_fatal_error`.

Reply:
Thinking about this more I'm coming around to the `Exhausted` suggestion. Maybe the enum/variants should be renamed to de-emphasize `Error` and emphasize path building continuation.

Something like:
```rust
enum BuildChainFailure {
    Rejected(Error),
    Exhausted(Error),
}
```

The intent being when we encounter a rejected chain path we can search for alternatives, but when we exhaust the budget we must stop.

Reply:
Oh, we could also reuse the std ControlFlow type? How does that feel?

Reply:
I think Rejected/Exhausted also does a pretty good job, BTW.

Reply:
> Oh, we could also reuse the std ControlFlow type? How does that feel?

That seems like a good idea. TIL.

Reply:
> Oh, we could also reuse the std ControlFlow type? How does that feel?

For this, the benefit might be that the ? operator does the right thing, which might be cool. (Not sure if that works today?)

Reply:
https://doc.rust-lang.org/core/ops/enum.ControlFlow.html -- also new to me and seems like a good fit!

Reply:
Switched to `ControlFlow` in 9fbc05a8668ccac7fd2123cd2bdb3462623cc64d

I put the `impl From<Error> for ControlFlow<Error, Error>` in `error.rs`, but there might be an argument to be made that it should be kept in `verify_cert.rs` where we're actually using error fatality to decide how to proceed with control flow.

WDYT?

Reply:
I would keep it close to Error.

Reply:
SGTM, going to call this conversation resolved :-) Thanks for the input (and teaching us about `ControlFlow`)